File: aragora/mcp/tools_module/integrations.py
"""
MCP Tools for External Integration triggers.

Provides tools for interacting with Zapier, Make, and n8n integrations:
- trigger_external_webhook: Trigger an external automation webhook
- list_integrations: List configured integrations
- test_integration: Test an integration connection
"""

from __future__ import annotations

import logging
import time
from typing import Any, Dict, List, Optional

logger = logging.getLogger(__name__)


async def trigger_external_webhook_tool(
    platform: str,
    event_type: str,
    data: str = "{}",
) -> Dict[str, Any]:
    """
    Trigger an external automation webhook (Zapier, Make, n8n).

    Args:
        platform: Target platform (zapier, make, n8n)
        event_type: Type of event to trigger
        data: JSON string of event data

    Returns:
        Dict with trigger results
    """
    import json as json_module

    valid_platforms = {"zapier", "make", "n8n"}
    if platform.lower() not in valid_platforms:
        return {"error": f"Invalid platform. Must be one of: {valid_platforms}"}

    # Parse event data
    try:
        event_data = json_module.loads(data) if data else {}
    except json_module.JSONDecodeError:
        return {"error": "Invalid JSON in data parameter"}

    # Add common event metadata
    event_data["timestamp"] = time.time()
    event_data["event_type"] = event_type
    event_data["source"] = "mcp_tool"

    try:
        triggered_count = 0

        if platform == "zapier":
            from aragora.integrations.zapier import get_zapier_integration

            zapier = get_zapier_integration()
            triggered_count = await zapier.fire_trigger(event_type, event_data)

        elif platform == "make":
            from aragora.integrations.make import get_make_integration

            make = get_make_integration()
            triggered_count = await make.trigger_webhooks(event_type, event_data)

        elif platform == "n8n":
            from aragora.integrations.n8n import get_n8n_integration

            n8n = get_n8n_integration()
            triggered_count = await n8n.dispatch_event(event_type, event_data)

        return {
            "platform": platform,
            "event_type": event_type,
            "triggered": triggered_count > 0,
            "webhooks_triggered": triggered_count,
        }

    except ImportError:
        logger.warning(f"{platform} integration not available")
        return {"error": f"{platform} integration module not available"}
    except Exception as e:
        logger.error(f"Failed to trigger {platform} webhook: {e}")
        return {"error": f"Trigger failed: {str(e)}"}


async def list_integrations_tool(
    platform: str = "all",
    workspace_id: Optional[str] = None,
) -> Dict[str, Any]:
    """
    List configured external integrations.

    Args:
        platform: Filter by platform (all, zapier, make, n8n)
        workspace_id: Filter by workspace ID

    Returns:
        Dict with list of configured integrations
    """
    integrations: Dict[str, List[Dict[str, Any]]] = {
        "zapier": [],
        "make": [],
        "n8n": [],
    }

    try:
        if platform in ("all", "zapier"):
            from aragora.integrations.zapier import get_zapier_integration

            zapier = get_zapier_integration()
            apps = zapier.list_apps(workspace_id)
            integrations["zapier"] = [
                {
                    "id": app.id,
                    "workspace_id": app.workspace_id,
                    "active": app.active,
                    "triggers_count": len(app.triggers),
                    "created_at": app.created_at,
                }
                for app in apps
            ]
    except ImportError:
        logger.debug("Zapier integration not available")
    except Exception as e:
        logger.warning(f"Failed to list Zapier integrations: {e}")

    try:
        if platform in ("all", "make"):
            from aragora.integrations.make import get_make_integration

            make = get_make_integration()
            connections = make.list_connections(workspace_id)
            integrations["make"] = [
                {
                    "id": conn.id,
                    "workspace_id": conn.workspace_id,
                    "active": conn.active,
                    "webhooks_count": len(conn.webhooks),
                    "created_at": conn.created_at,
                }
                for conn in connections
            ]
    except ImportError:
        logger.debug("Make integration not available")
    except Exception as e:
        logger.warning(f"Failed to list Make integrations: {e}")

    try:
        if platform in ("all", "n8n"):
            from aragora.integrations.n8n import get_n8n_integration

            n8n = get_n8n_integration()
            credentials = n8n.list_credentials(workspace_id)
            integrations["n8n"] = [
                {
                    "id": cred.id,
                    "workspace_id": cred.workspace_id,
                    "active": cred.active,
                    "webhooks_count": len(cred.webhooks),
                    "created_at": cred.created_at,
                }
                for cred in credentials
            ]
    except ImportError:
        logger.debug("n8n integration not available")
    except Exception as e:
        logger.warning(f"Failed to list n8n integrations: {e}")

    # Calculate totals
    total = sum(len(v) for v in integrations.values())

    return {
        "integrations": integrations if platform == "all" else {platform: integrations.get(platform, [])},
        "total": total,
        "platform_filter": platform,
        "workspace_filter": workspace_id,
    }


async def test_integration_tool(
    platform: str,
    integration_id: str,
) -> Dict[str, Any]:
    """
    Test an integration connection.

    Args:
        platform: Platform (zapier, make, n8n)
        integration_id: ID of the integration to test

    Returns:
        Dict with test results
    """
    valid_platforms = {"zapier", "make", "n8n"}
    if platform.lower() not in valid_platforms:
        return {"error": f"Invalid platform. Must be one of: {valid_platforms}"}

    try:
        if platform == "zapier":
            from aragora.integrations.zapier import get_zapier_integration

            zapier = get_zapier_integration()
            app = zapier.get_app(integration_id)
            if not app:
                return {"error": f"Zapier app {integration_id} not found"}
            return {
                "platform": "zapier",
                "integration_id": integration_id,
                "status": "ok" if app.active else "inactive",
                "triggers_configured": len(app.triggers),
                "total_triggers_fired": app.trigger_count,
            }

        elif platform == "make":
            from aragora.integrations.make import get_make_integration

            make = get_make_integration()
            result = make.test_connection(integration_id)
            return {
                "platform": "make",
                "integration_id": integration_id,
                **result,
            }

        elif platform == "n8n":
            from aragora.integrations.n8n import get_n8n_integration

            n8n = get_n8n_integration()
            cred = n8n.get_credential(integration_id)
            if not cred:
                return {"error": f"n8n credential {integration_id} not found"}
            return {
                "platform": "n8n",
                "integration_id": integration_id,
                "status": "ok" if cred.active else "inactive",
                "webhooks_configured": len(cred.webhooks),
                "total_operations": cred.operation_count,
            }

    except ImportError:
        return {"error": f"{platform} integration module not available"}
    except Exception as e:
        logger.error(f"Failed to test {platform} integration: {e}")
        return {"error": f"Test failed: {str(e)}"}

    return {"error": "Unknown platform"}


async def get_integration_events_tool(
    platform: str,
) -> Dict[str, Any]:
    """
    Get available event types for an integration platform.

    Args:
        platform: Platform (zapier, make, n8n)

    Returns:
        Dict with available event types
    """
    valid_platforms = {"zapier", "make", "n8n"}
    if platform.lower() not in valid_platforms:
        return {"error": f"Invalid platform. Must be one of: {valid_platforms}"}

    try:
        if platform == "zapier":
            from aragora.integrations.zapier import get_zapier_integration

            zapier = get_zapier_integration()
            return {
                "platform": "zapier",
                "trigger_types": zapier.TRIGGER_TYPES,
                "action_types": zapier.ACTION_TYPES,
            }

        elif platform == "make":
            from aragora.integrations.make import get_make_integration

            make = get_make_integration()
            modules = make.MODULE_TYPES
            triggers = {k: v for k, v in modules.items() if v.get("type") == "trigger"}
            actions = {k: v for k, v in modules.items() if v.get("type") == "action"}
            return {
                "platform": "make",
                "trigger_modules": triggers,
                "action_modules": actions,
            }

        elif platform == "n8n":
            from aragora.integrations.n8n import get_n8n_integration

            n8n = get_n8n_integration()
            return {
                "platform": "n8n",
                "event_types": n8n.EVENT_TYPES,
            }

    except ImportError:
        return {"error": f"{platform} integration module not available"}
    except Exception as e:
        logger.error(f"Failed to get {platform} events: {e}")
        return {"error": f"Failed: {str(e)}"}

    return {"error": "Unknown platform"}


# Export all tools
__all__ = [
    "trigger_external_webhook_tool",
    "list_integrations_tool",
    "test_integration_tool",
    "get_integration_events_tool",
]

File: sdk/typescript/src/namespaces/integrations.ts
/**
 * Integrations Namespace API
 *
 * Provides a namespaced interface for external integration management.
 * Essential for connecting Aragora to external services.
 */

/**
 * Integration definition
 */
export interface Integration {
  id: string;
  type: string;
  name: string;
  description?: string;
  status: 'connected' | 'disconnected' | 'error' | 'pending';
  config: Record<string, unknown>;
  credentials_stored: boolean;
  last_sync_at?: string;
  error_message?: string;
  created_at: string;
  updated_at: string;
}

/**
 * Available integration type
 */
export interface AvailableIntegration {
  type: string;
  name: string;
  description: string;
  category: 'communication' | 'storage' | 'productivity' | 'analytics' | 'custom';
  icon_url?: string;
  oauth_supported: boolean;
  required_fields: string[];
  optional_fields: string[];
  documentation_url?: string;
  setup_time_minutes: number;
}

/**
 * Integration config schema
 */
export interface IntegrationConfigSchema {
  type: string;
  required_fields: {
    name: string;
    type: string;
    description: string;
    sensitive?: boolean;
  }[];
  optional_fields: {
    name: string;
    type: string;
    description: string;
    default?: unknown;
  }[];
}

/**
 * Integration sync status
 */
export interface IntegrationSyncStatus {
  integration_id: string;
  last_sync_at?: string;
  next_sync_at?: string;
  sync_frequency_minutes: number;
  items_synced: number;
  errors: string[];
  status: 'idle' | 'syncing' | 'error';
}

/**
 * Create integration request
 */
export interface CreateIntegrationRequest {
  type: string;
  name?: string;
  config: Record<string, unknown>;
}

/**
 * Update integration request
 */
export interface UpdateIntegrationRequest {
  name?: string;
  config?: Record<string, unknown>;
  enabled?: boolean;
}

/**
 * Bot status response
 */
export interface BotStatus {
  platform: string;
  connected: boolean;
  last_activity_at?: string;
  active_debates: number;
  error?: string;
}

/**
 * Zapier app summary (from list)
 */
export interface ZapierAppSummary {
  id: string;
  workspace_id: string;
  created_at: string;
  active: boolean;
  trigger_count: number;
  action_count: number;
}

/**
 * Zapier app with credentials (from create)
 */
export interface ZapierApp {
  id: string;
  workspace_id: string;
  api_key: string;
  api_secret: string;
  created_at: string;
}

/**
 * Zapier trigger subscription
 */
export interface ZapierTriggerSubscription {
  id: string;
  trigger_type: string;
  webhook_url: string;
  created_at: string;
}

/**
 * Zapier trigger type definition
 */
export interface ZapierTrigger {
  id: string;
  name: string;
  description: string;
  sample_payload: Record<string, unknown>;
}

/**
 * Create Zapier app request
 */
export interface CreateZapierAppRequest {
  workspace_id: string;
}

/**
 * Subscribe to Zapier trigger request
 */
export interface SubscribeZapierTriggerRequest {
  app_id: string;
  trigger_type: string;
  webhook_url: string;
  workspace_id?: string;
  debate_tags?: string[];
  min_confidence?: number;
}

/**
 * Make connection summary (from list)
 */
export interface MakeConnectionSummary {
  id: string;
  workspace_id: string;
  created_at: string;
  active: boolean;
  total_operations: number;
  webhooks_count: number;
}

/**
 * Make connection with credentials (from create)
 */
export interface MakeConnection {
  id: string;
  workspace_id: string;
  api_key: string;
  created_at: string;
}

/**
 * Make webhook
 */
export interface MakeWebhook {
  id: string;
  module_type: string;
  webhook_url: string;
  created_at: string;
}

/**
 * Make module definition
 */
export interface MakeModule {
  id: string;
  name: string;
  description: string;
  type: 'action' | 'trigger' | 'search';
  inputs: { name: string; type: string; required: boolean }[];
  outputs: { name: string; type: string }[];
}

/**
 * Create Make connection request
 */
export interface CreateMakeConnectionRequest {
  workspace_id: string;
}

/**
 * Register Make webhook request
 */
export interface RegisterMakeWebhookRequest {
  connection_id: string;
  module_type: string;
  webhook_url: string;
  workspace_id?: string;
  event_filter?: Record<string, unknown>;
}

/**
 * n8n credential summary (from list)
 */
export interface N8nCredentialSummary {
  id: string;
  workspace_id: string;
  api_url: string;
  created_at: string;
  active: boolean;
  operation_count: number;
  webhooks_count: number;
}

/**
 * n8n credential with API key (from create)
 */
export interface N8nCredential {
  id: string;
  workspace_id: string;
  api_key: string;
  api_url: string;
  created_at: string;
}

/**
 * n8n webhook
 */
export interface N8nWebhook {
  id: string;
  webhook_path: string;
  events: string[];
  created_at: string;
}

/**
 * n8n node definition
 */
export interface N8nNode {
  name: string;
  displayName: string;
  description: string;
  version: number;
  inputs: string[];
  outputs: string[];
  properties: Record<string, unknown>[];
}

/**
 * n8n nodes response (includes node, trigger, credential definitions)
 */
export interface N8nNodesResponse {
  node: N8nNode;
  trigger: N8nNode;
  credential: Record<string, unknown>;
  events: Record<string, unknown>;
}

/**
 * Create n8n credential request
 */
export interface CreateN8nCredentialRequest {
  workspace_id: string;
  api_url?: string;
}

/**
 * Register n8n webhook request
 */
export interface RegisterN8nWebhookRequest {
  credential_id: string;
  events: string[];
  workflow_id?: string;
  node_id?: string;
  workspace_id?: string;
}

/**
 * Teams status
 */
export interface TeamsStatus {
  connected: boolean;
  tenant_id?: string;
  bot_id?: string;
  channels: number;
  error?: string;
}

/**
 * Wizard provider
 */
export interface WizardProvider {
  type: string;
  name: string;
  description: string;
  category: string;
  icon_url?: string;
  oauth_url?: string;
  fields: { name: string; type: string; required: boolean; description: string }[];
}

/**
 * Wizard step status
 */
export interface WizardStepStatus {
  step: string;
  status: 'pending' | 'in_progress' | 'completed' | 'failed';
  message?: string;
  error?: string;
}

/**
 * Integration health status
 */
export interface IntegrationHealth {
  type: string;
  status: 'healthy' | 'degraded' | 'unhealthy';
  latency_ms?: number;
  last_check_at: string;
  details?: Record<string, unknown>;
}

/**
 * Integration stats
 */
export interface IntegrationStats {
  total_integrations: number;
  active_integrations: number;
  by_type: Record<string, number>;
  by_status: Record<string, number>;
  total_syncs_24h: number;
  failed_syncs_24h: number;
}

/**
 * Interface for the internal client used by IntegrationsAPI.
 */
interface IntegrationsClientInterface {
  get<T>(path: string): Promise<T>;
  post<T>(path: string, body?: unknown): Promise<T>;
  put<T>(path: string, body?: unknown): Promise<T>;
  delete<T>(path: string): Promise<T>;
  request<T = unknown>(method: string, path: string, options?: { params?: Record<string, unknown>; json?: Record<string, unknown> }): Promise<T>;
}

/**
 * Integrations API namespace.
 *
 * Provides methods for managing external integrations:
 * - Connect to external services (Slack, Teams, etc.)
 * - Configure integration settings
 * - Monitor sync status
 * - Test connections
 *
 * @example
 * ```typescript
 * const client = createClient({ baseUrl: 'https://api.aragora.ai', apiKey: 'your-key' });
 *
 * // List available integrations
 * const { integrations: available } = await client.integrations.listAvailable();
 *
 * // Get config schema for Slack
 * const schema = await client.integrations.getConfigSchema('slack');
 *
 * // Create a Slack integration
 * const integration = await client.integrations.create({
 *   type: 'slack',
 *   config: { workspace_id: 'T123456', channel: '#general' },
 * });
 *
 * // Test the connection
 * const result = await client.integrations.test(integration.id);
 *
 * // Trigger a sync
 * const status = await client.integrations.sync(integration.id);
 * ```
 */
export class IntegrationsAPI {
  constructor(private client: IntegrationsClientInterface) {}

  // ===========================================================================
  // Integration Discovery
  // ===========================================================================

  /**
   * List available integration types.
   * @route GET /api/v1/integrations/available
   */
  async listAvailable(): Promise<{ integrations: AvailableIntegration[] }> {
    return this.client.request('GET', '/api/v1/integrations/available') as Promise<{ integrations: AvailableIntegration[] }>;
  }

  /**
   * Get integration status summary.
   * @route GET /api/v1/integrations/status
   */
  async getStatus(): Promise<Record<string, unknown>> {
    return this.client.request('GET', '/api/v1/integrations/status') as Promise<Record<string, unknown>>;
  }

  /**
   * Get global integration configuration.
   * @route GET /api/v1/integrations/config
   */
  async getConfig(): Promise<Record<string, unknown>> {
    return this.client.request('GET', '/api/v1/integrations/config') as Promise<Record<string, unknown>>;
  }

  /**
   * Get configuration schema for an integration type.
   * @route GET /api/v1/integrations/config/{type}
   */
  async getConfigSchema(type: string): Promise<IntegrationConfigSchema> {
    return this.client.request('GET', `/api/v1/integrations/config/${type}`) as Promise<IntegrationConfigSchema>;
  }

  // ===========================================================================
  // Integration Management
  // ===========================================================================

  /**
   * List all configured integrations.
   */
  async list(): Promise<{ integrations: Integration[]; total: number }> {
    return this.client.get('/api/integrations');
  }

  /**
   * Get a specific integration by ID.
   */
  async get(id: string): Promise<Integration> {
    return this.client.get(`/api/integrations/${id}`);
  }

  /**
   * Create a new integration.
   */
  async create(body: CreateIntegrationRequest): Promise<Integration> {
    return this.client.post('/api/integrations', body);
  }

  /**
   * Update an existing integration.
   */
  async update(id: string, body: UpdateIntegrationRequest): Promise<Integration> {
    return this.client.put(`/api/integrations/${id}`, body);
  }

  /**
   * Delete an integration.
   */
  async delete(id: string): Promise<{ deleted: boolean }> {
    return this.client.delete(`/api/integrations/${id}`);
  }

  // ===========================================================================
  // Testing and Sync
  // ===========================================================================

  /**
   * Test an integration connection.
   */
  async test(id: string): Promise<{ success: boolean; message?: string; error?: string }> {
    return this.client.post(`/api/integrations/${id}/test`);
  }

  /**
   * Trigger a sync for an integration.
   */
  async sync(id: string): Promise<IntegrationSyncStatus> {
    return this.client.post(`/api/integrations/${id}/sync`);
  }

  /**
   * Get sync status for an integration.
   */
  async getSyncStatus(id: string): Promise<IntegrationSyncStatus> {
    return this.client.get(`/api/integrations/${id}/sync`);
  }

  // ===========================================================================
  // Bot Platforms
  // ===========================================================================

  /**
   * Get Slack bot status.
   */
  async getSlackStatus(): Promise<BotStatus> {
    return this.client.request('GET', '/api/v1/bots/slack/status');
  }

  /**
   * Get Telegram bot status.
   */
  async getTelegramStatus(): Promise<BotStatus> {
    return this.client.request('GET', '/api/v1/bots/telegram/status');
  }

  /**
   * Get WhatsApp bot status.
   */
  async getWhatsAppStatus(): Promise<BotStatus> {
    return this.client.request('GET', '/api/v1/bots/whatsapp/status');
  }

  /**
   * Get Discord bot status.
   */
  async getDiscordStatus(): Promise<BotStatus> {
    return this.client.request('GET', '/api/v1/bots/discord/status');
  }

  /**
   * Get Google Chat bot status.
   */
  async getGoogleChatStatus(): Promise<BotStatus> {
    return this.client.request('GET', '/api/v1/bots/google-chat/status');
  }

  /**
   * Get email webhook status.
   */
  async getEmailStatus(): Promise<{ sendgrid: BotStatus; ses: BotStatus }> {
    return this.client.request('GET', '/api/v1/bots/email/status');
  }

  // ===========================================================================
  // Teams Integration
  // ===========================================================================

  /**
   * Get Microsoft Teams integration status.
   */
  async getTeamsStatus(): Promise<TeamsStatus> {
    return this.client.request('GET', '/api/v1/integrations/teams/status');
  }

  /**
   * Install Microsoft Teams app.
   */
  async installTeams(tenantId: string): Promise<{ installed: boolean; bot_id: string }> {
    return this.client.request('POST', '/api/integrations/teams/install', { json: { tenant_id: tenantId } });
  }

  /**
   * Handle Teams OAuth callback.
   */
  async teamsCallback(code: string, state: string): Promise<{ success: boolean }> {
    return this.client.request('POST', '/api/integrations/teams/callback', { json: { code, state } });
  }

  /**
   * Refresh Teams token.
   */
  async refreshTeamsToken(): Promise<{ refreshed: boolean }> {
    return this.client.request('POST', '/api/integrations/teams/refresh');
  }

  /**
   * Send notification to Teams channel.
   */
  async notifyTeams(channelId: string, message: string, options?: { card?: Record<string, unknown> }): Promise<{ sent: boolean; message_id: string }> {
    return this.client.request('POST', '/api/v1/integrations/teams/notify', { json: { channel_id: channelId, message, ...options } });
  }

  // ===========================================================================
  // Discord Integration
  // ===========================================================================

  /**
   * Install Discord bot.
   */
  async installDiscord(guildId: string): Promise<{ installed: boolean; bot_id: string }> {
    return this.client.request('POST', '/api/integrations/discord/install', { json: { guild_id: guildId } });
  }

  /**
   * Handle Discord OAuth callback.
   */
  async discordCallback(code: string, state: string): Promise<{ success: boolean }> {
    return this.client.request('POST', '/api/integrations/discord/callback', { json: { code, state } });
  }

  /**
   * Uninstall Discord bot.
   */
  async uninstallDiscord(guildId: string): Promise<{ uninstalled: boolean }> {
    return this.client.request('POST', '/api/integrations/discord/uninstall', { json: { guild_id: guildId } });
  }

  // ===========================================================================
  // Zapier Integration
  // ===========================================================================

  /**
   * List Zapier apps.
   *
   * @param workspaceId - Optional workspace ID to filter by
   * @returns List of Zapier apps
   */
  async listZapierApps(workspaceId?: string): Promise<{ apps: ZapierAppSummary[]; count: number }> {
    const params = workspaceId ? { params: { workspace_id: workspaceId } } : undefined;
    return this.client.request('GET', '/api/v1/integrations/zapier/apps', params);
  }

  /**
   * Create Zapier app.
   *
   * @param body - Create request with workspace_id
   * @returns Created app with API credentials (save these, they won't be shown again)
   */
  async createZapierApp(body: CreateZapierAppRequest): Promise<{ app: ZapierApp; message: string }> {
    return this.client.request('POST', '/api/v1/integrations/zapier/apps', { json: { ...body } });
  }

  /**
   * Delete Zapier app.
   *
   * @param appId - The app ID to delete
   * @returns Deletion confirmation
   */
  async deleteZapierApp(appId: string): Promise<{ deleted: boolean; app_id: string }> {
    return this.client.request('DELETE', `/api/v1/integrations/zapier/apps/${appId}`);
  }

  /**
   * List available Zapier triggers and actions.
   *
   * @returns Available trigger and action types
   */
  async listZapierTriggerTypes(): Promise<{ triggers: Record<string, unknown>; actions: Record<string, unknown> }> {
    return this.client.request('GET', '/api/v1/integrations/zapier/triggers');
  }

  /**
   * Subscribe to Zapier trigger.
   *
   * @param body - Subscribe request with app_id, trigger_type, webhook_url
   * @returns Created trigger subscription
   */
  async subscribeZapierTrigger(body: SubscribeZapierTriggerRequest): Promise<{ trigger: ZapierTriggerSubscription }> {
    return this.client.request('POST', '/api/v1/integrations/zapier/triggers', { json: { ...body } });
  }

  /**
   * Unsubscribe from Zapier trigger.
   *
   * @param triggerId - The trigger subscription ID
   * @param appId - The app ID the trigger belongs to
   * @returns Deletion confirmation
   */
  async unsubscribeZapierTrigger(triggerId: string, appId: string): Promise<{ deleted: boolean; trigger_id: string }> {
    return this.client.request('DELETE', `/api/v1/integrations/zapier/triggers/${triggerId}`, { params: { app_id: appId } });
  }

  // ===========================================================================
  // Make (Integromat) Integration
  // ===========================================================================

  /**
   * List Make connections.
   *
   * @param workspaceId - Optional workspace ID to filter by
   * @returns List of Make connections
   */
  async listMakeConnections(workspaceId?: string): Promise<{ connections: MakeConnectionSummary[]; count: number }> {
    const params = workspaceId ? { params: { workspace_id: workspaceId } } : undefined;
    return this.client.request('GET', '/api/v1/integrations/make/connections', params);
  }

  /**
   * Create Make connection.
   *
   * @param body - Create request with workspace_id
   * @returns Created connection with API key (save this, it won't be shown again)
   */
  async createMakeConnection(body: CreateMakeConnectionRequest): Promise<{ connection: MakeConnection; message: string }> {
    return this.client.request('POST', '/api/v1/integrations/make/connections', { json: { ...body } });
  }

  /**
   * Delete Make connection.
   *
   * @param connectionId - The connection ID to delete
   * @returns Deletion confirmation
   */
  async deleteMakeConnection(connectionId: string): Promise<{ deleted: boolean; connection_id: string }> {
    return this.client.request('DELETE', `/api/v1/integrations/make/connections/${connectionId}`);
  }

  /**
   * List Make modules.
   *
   * @returns Available module types
   */
  async listMakeModules(): Promise<{ modules: Record<string, unknown> }> {
    return this.client.request('GET', '/api/v1/integrations/make/modules');
  }

  /**
   * Register Make webhook.
   *
   * @param body - Register request with connection_id, module_type, webhook_url
   * @returns Created webhook
   */
  async registerMakeWebhook(body: RegisterMakeWebhookRequest): Promise<{ webhook: MakeWebhook }> {
    return this.client.request('POST', '/api/v1/integrations/make/webhooks', { json: { ...body } });
  }

  /**
   * Unregister Make webhook.
   *
   * @param webhookId - The webhook ID to unregister
   * @param connectionId - The connection ID the webhook belongs to
   * @returns Deletion confirmation
   */
  async unregisterMakeWebhook(webhookId: string, connectionId: string): Promise<{ deleted: boolean; webhook_id: string }> {
    return this.client.request('DELETE', `/api/v1/integrations/make/webhooks/${webhookId}`, { params: { connection_id: connectionId } });
  }

  // ===========================================================================
  // n8n Integration
  // ===========================================================================

  /**
   * List n8n credentials.
   *
   * @param workspaceId - Optional workspace ID to filter by
   * @returns List of n8n credentials
   */
  async listN8nCredentials(workspaceId?: string): Promise<{ credentials: N8nCredentialSummary[]; count: number }> {
    const params = workspaceId ? { params: { workspace_id: workspaceId } } : undefined;
    return this.client.request('GET', '/api/v1/integrations/n8n/credentials', params);
  }

  /**
   * Create n8n credential.
   *
   * @param body - Create request with workspace_id and optional api_url
   * @returns Created credential with API key (save this, it won't be shown again)
   */
  async createN8nCredential(body: CreateN8nCredentialRequest): Promise<{ credential: N8nCredential; message: string }> {
    return this.client.request('POST', '/api/v1/integrations/n8n/credentials', { json: { ...body } });
  }

  /**
   * Delete n8n credential.
   *
   * @param credentialId - The credential ID to delete
   * @returns Deletion confirmation
   */
  async deleteN8nCredential(credentialId: string): Promise<{ deleted: boolean; credential_id: string }> {
    return this.client.request('DELETE', `/api/v1/integrations/n8n/credentials/${credentialId}`);
  }

  /**
   * Get n8n node definitions.
   *
   * @returns Node, trigger, and credential definitions plus event types
   */
  async getN8nNodes(): Promise<N8nNodesResponse> {
    return this.client.request('GET', '/api/v1/integrations/n8n/nodes');
  }

  /**
   * Register n8n webhook.
   *
   * @param body - Register request with credential_id and events
   * @returns Created webhook with webhook_path
   */
  async registerN8nWebhook(body: RegisterN8nWebhookRequest): Promise<{ webhook: N8nWebhook }> {
    return this.client.request('POST', '/api/v1/integrations/n8n/webhooks', { json: { ...body } });
  }

  /**
   * Unregister n8n webhook.
   *
   * @param webhookId - The webhook ID to unregister
   * @param credentialId - The credential ID the webhook belongs to
   * @returns Deletion confirmation
   */
  async unregisterN8nWebhook(webhookId: string, credentialId: string): Promise<{ deleted: boolean; webhook_id: string }> {
    return this.client.request('DELETE', `/api/v1/integrations/n8n/webhooks/${webhookId}`, { params: { credential_id: credentialId } });
  }

  // ===========================================================================
  // Integration Wizard (v2)
  // ===========================================================================

  /**
   * Start integration wizard.
   */
  async startWizard(integrationType: string): Promise<{ session_id: string; steps: string[] }> {
    return this.client.request('POST', '/api/v2/integrations/wizard', { json: { type: integrationType } });
  }

  /**
   * List wizard providers.
   */
  async listWizardProviders(): Promise<{ providers: WizardProvider[] }> {
    return this.client.request('GET', '/api/v2/integrations/wizard/providers');
  }

  /**
   * Get wizard status.
   */
  async getWizardStatus(sessionId: string): Promise<{ steps: WizardStepStatus[]; current_step: string }> {
    return this.client.request('GET', '/api/v2/integrations/wizard/status', { params: { session_id: sessionId } });
  }

  /**
   * Validate wizard step.
   */
  async validateWizardStep(sessionId: string, step: string, data: Record<string, unknown>): Promise<{ valid: boolean; errors?: string[] }> {
    return this.client.request('POST', '/api/v2/integrations/wizard/validate', { json: { session_id: sessionId, step, data } });
  }

  // ===========================================================================
  // v2 Integration Management
  // ===========================================================================

  /**
   * List integrations (v2).
   */
  async listV2(options?: { type?: string; status?: string; limit?: number; offset?: number }): Promise<{ integrations: Integration[]; total: number }> {
    return this.client.request('GET', '/api/v2/integrations', { params: options });
  }

  /**
   * Get integration by type (v2).
   */
  async getByType(type: string): Promise<Integration> {
    return this.client.request('GET', `/api/v2/integrations/${type}`);
  }

  /**
   * Get integration health by type.
   */
  async getHealth(type: string): Promise<IntegrationHealth> {
    return this.client.request('GET', `/api/v2/integrations/${type}/health`);
  }

  /**
   * Test integration by type.
   */
  async testByType(type: string): Promise<{ success: boolean; latency_ms?: number; error?: string }> {
    return this.client.request('POST', `/api/v2/integrations/${type}/test`);
  }

  /**
   * Get integration stats.
   */
  async getStats(): Promise<IntegrationStats> {
    return this.client.request('GET', '/api/v2/integrations/stats');
  }

  /**
   * Test any platform integration.
   */
  async testPlatform(platform: string): Promise<{ success: boolean; error?: string }> {
    return this.client.request('POST', `/api/v1/integrations/${platform}/test`);
  }
}
